theta_dup/include/theta_sketch_dup_impl.hpp

```diff
@@ -0,0 +1,1045 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef theta_sketch_dup_DUP_IMPL_HPP_
+#define theta_sketch_dup_DUP_IMPL_HPP_
+
+#include <algorithm>
+#include <cmath>
+#include <memory>
+#include <functional>
+#include <istream>
+#include <ostream>
+#include <sstream>
+
+#include "MurmurHash3.h"
+#include "serde.hpp"
+#include "binomial_bounds.hpp"
+#include "memory_operations.hpp"
+
+namespace datasketches {
+
+/*
+ * author Alexander Saydakov
+ * author Lee Rhodes
+ * author Kevin Lang
+ */
+
+template<typename A>
+theta_sketch_dup_alloc<A>::theta_sketch_dup_alloc(bool is_empty, uint64_t theta):
+is_empty_(is_empty), theta_(theta)
+{}
+
+template<typename A>
+bool theta_sketch_dup_alloc<A>::is_empty() const {
+  return is_empty_;
+}
+
+template<typename A>
+double theta_sketch_dup_alloc<A>::get_estimate() const {
+  return get_num_retained() / get_theta();
+}
+
+template<typename A>
+double theta_sketch_dup_alloc<A>::get_lower_bound(uint8_t num_std_devs) const {
+  if (!is_estimation_mode()) return get_num_retained();
+  return binomial_bounds::get_lower_bound(get_num_retained(), get_theta(), num_std_devs);
+}
+
+template<typename A>
+double theta_sketch_dup_alloc<A>::get_upper_bound(uint8_t num_std_devs) const {
+  if (!is_estimation_mode()) return get_num_retained();
+  return binomial_bounds::get_upper_bound(get_num_retained(), get_theta(), num_std_devs);
+}
+
+template<typename A>
+bool theta_sketch_dup_alloc<A>::is_estimation_mode() const {
+  return theta_ < MAX_THETA && !is_empty_;
+}
+
+template<typename A>
+double theta_sketch_dup_alloc<A>::get_theta() const {
+  return (double) theta_ / MAX_THETA;
+}
+
+template<typename A>
+uint64_t theta_sketch_dup_alloc<A>::get_theta64() const {
+  return theta_;
+}
+
+template<typename A>
```

Comment:
Could you keep the function def in header file and put the implementation in a .cc file? Also add comments/description for each function in the header file. Thanks :)

Reply:
Sure!

Reply:
This header file uses **template class**, which is not usually compiled separately, unless we put explicitly all required template class instantiation in a .cc file [(link to google c++ style guide)](https://google.github.io/styleguide/cppguide.html#Self_contained_Headers). There are two ways of writing our header files according to the google c++ style guide:

1. (suggested way)  placing the definitions for template and inline functions in the same file as their declarations.
2. Non-header files that are meant for inclusion should end in .inc and be used sparingly.